### PR TITLE
[SPARK-15704] [SQL] add a test case in DatasetAggregatorSuite for regression testing 

### PR DESCRIPTION
**Pull request by inouehrs**
## What changes were proposed in this pull request?

This change fixes a crash in TungstenAggregate while executing "Dataset complex Aggregator" test case due to IndexOutOfBoundsException.

jira entry for detail: https://issues.apache.org/jira/browse/SPARK-15704
## How was this patch tested?

Using existing unit tests (including DatasetBenchmark)

### CODE REVIEW

**Comment by rxin** [comment]
cc @cloud-fan / @sameeragarwal

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala [hunk 1]
```diff
@@ -34,6 +34,11 @@ object TypedAggregateExpression {
     val bufferEncoder = encoderFor[BUF]
     val bufferSerializer = bufferEncoder.namedExpressions
     val bufferDeserializer = bufferEncoder.deserializer.transform {
+      case UnresolvedAttribute(nameParts) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala]: good catch! can you try if we can fix it in this way? https://github.com/apache/spark/pull/13269/files#diff-585631b931d2b881e65c585dec11cadbR36

I think it's less hacky

**Comment by rxin** [comment]
Thanks - can you also update the title? TangstenAggregate -> TungstenAggregate

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala (hunk 1 above)
Comment (inouehrs) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala]: It seems that the new cleaner implementation does not need my change; it handles unresolved attributes correctly. Thank you!

**Comment by inouehrs** [comment]
I fixed my typos.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala]: we can apply that part of code in this PR and fix this bug first

**Comment by inouehrs** [comment]
I rewrote the patch as suggested by @cloud-fan (based on https://github.com/apache/spark/pull/13269/files#diff-585631b931d2b881e65c585dec11cadbR36).
Also, I added a test case to reproduce the error in DatasetAggregatorSuite.

**Comment by inouehrs** [comment]
Since the new encoder framework ([SPARK-15494]) has been merged, my bug fix part is no longer needed.
I appreciate if the new test case part can be merged. Thanks!

**Comment by cloud-fan** [comment]
can you revert the file mode change? the test LGTM

**Comment by inouehrs** [comment]
I reverted the file mode change. Thank you!

**Comment by cloud-fan** [comment]
can you also update your title to say that this PR only adds regression test? thanks!

**Comment by inouehrs** [comment]
TItle changed. Thanks

**Comment by rxin** [comment]
@cloud-fan next time please leave a message on the pr saying it was merged and the branches it was merged in.

**Comment by cloud-fan** [comment]
Sorry I was interrupted by something and forgot about it...
thanks, merging to master and 2.0!

**Pull request closed without merging**